src/details/ArborX_DetailsBufferOptimization.hpp

```diff
@@ -0,0 +1,230 @@
+/****************************************************************************
+ * Copyright (c) 2012-2020 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+#ifndef ARBORX_DETAILS_BUFFER_OPTIMIZATON_HPP
+#define ARBORX_DETAILS_BUFFER_OPTIMIZATON_HPP
+
+#include <ArborX_DetailsUtils.hpp>
+#include <ArborX_Exception.hpp>
+#include <ArborX_Macros.hpp>
+#include <ArborX_Traits.hpp>
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX
+{
+namespace Details
+{
+
+struct FirstPassTag
+{
+};
+struct FirstPassNoBufferOptimizationTag
+{
+};
+struct SecondPassTag
+{
+};
+
+template <typename Tag, typename Predicates, typename Callback,
+          typename OutputView, typename CountView, typename OffsetView>
+struct WrappedCallback;
+
+template <typename Predicates, typename Callback, typename OutputView,
+          typename CountView, typename OffsetView>
+struct WrappedCallback<FirstPassTag, Predicates, Callback, OutputView,
+                       CountView, OffsetView>
+{
+  Predicates predicates_;
+  Callback callback_;
+  OutputView out_;
+  CountView counts_;
+  OffsetView offset_;
+  int j;
+
+  using ValueType = typename OutputView::value_type;
+  using Access = Traits::Access<Predicates, Traits::PredicatesTag>;
+
+  KOKKOS_FUNCTION void operator()(int i) const
+  {
+    if (offset_(j) + counts_(j) < offset_(j + 1))
+      callback_(Access::get(predicates_, j), i, [&](ValueType const &value) {
+        out_(offset_(j) + Kokkos::atomic_fetch_add(&counts_(j), 1)) = value;
+      });
+    else
+      callback_(Access::get(predicates_, j), i, [&](ValueType const &) {
+        Kokkos::atomic_fetch_add(&counts_(j), 1);
+      });
+  }
+};
+
+template <typename Predicates, typename Callback, typename OutputView,
+          typename CountView, typename OffsetView>
+struct WrappedCallback<FirstPassNoBufferOptimizationTag, Predicates, Callback,
+                       OutputView, CountView, OffsetView>
+{
+  Predicates predicates_;
+  Callback callback_;
+  OutputView out_;
+  CountView counts_;
+  OffsetView offset_;
+  int j;
+
+  using ValueType = typename OutputView::value_type;
+  using Access = Traits::Access<Predicates, Traits::PredicatesTag>;
+
+  KOKKOS_FUNCTION void operator()(int i) const
+  {
+    callback_(Access::get(predicates_, j), i, [&](ValueType const &) {
+      Kokkos::atomic_fetch_add(&counts_(j), 1);
```

Comment:
Just an observation: MAP shows this line getting hit pretty hard, taking about 13% of radius search runtime in Serial.

Reply:
I think this may similar to the one we've [seen](https://github.com/arborx/ArborX/pull/166#issuecomment-578893224) before, that was fixed by e25f8057f6fabdf1135a276108cfd8aa59ad3aaa. It may be that Serial is slowed down due to atomics, while Cuda due to shared array.

Reply:
I don't follow what `i` and `j` here are. This seems to be called from within `WrappedBVH`, but from there it seems `i` is query index. But in this code it looks like primitive index. I also don't see where `j` is set. Maybe I should go to sleep.

Reply:
`j`is predicate, `i` is primitive.